air/src/proof/ood_frame.rs

```diff
@@ -38,18 +37,32 @@ impl OodFrame {
     // UPDATERS
     // --------------------------------------------------------------------------------------------
 
-    /// Updates the trace state portion of this out-of-domain frame.
+    /// Updates the trace state portion of this out-of-domain frame. This also returns a compactified
+    /// version of the out-of-domain frame with the rows interleaved. This is done so that reseeding
+    /// of the random coin needs to be done only once as opposed to once per each row.
     ///
     /// # Panics
     /// Panics if evaluation frame has already been set.
-    pub fn set_trace_states<E: FieldElement>(&mut self, trace_states: &[Vec<E>]) {
+    pub fn set_trace_states<E: FieldElement>(&mut self, trace_states: &[Vec<E>]) -> Vec<E> {
         assert!(
             self.trace_states.is_empty(),
             "trace sates have already been set"
         );
-        for trace_state in trace_states {
-            trace_state.write_into(&mut self.trace_states);
+
+        // save the evaluations with the current and next evaluations interleaved for each polynomial
+        let frame_size = trace_states.len();
+        let width = trace_states[0].len();
+
+        let mut result = vec![];
+        for i in 0..width {
+            for row in trace_states.iter() {
+                result.push(row[i]);
+            }
         }
+        self.trace_states.write_u8(frame_size as u8);
```

Comment:
nit: should we put a `debug_assert` here to check that the frame size can fit into u8? Maybe something like:
```Rust
debug_assert!(frame_size <= u8::MAX as usize);
```

Reply:
Done!

Reply:
Just a curiosity: What is the best way, in your opinion, to use newlines to make the following code more readable?

```rust

        let frame_size = trace_states.len();
        let width = trace_states[0].len();

        let mut result = vec![];
        for i in 0..width {
            for row in trace_states.iter() {
                result.push(row[i]);
            }
        }
        debug_assert!(frame_size <= u8::MAX as usize);
        self.trace_states.write_u8(frame_size as u8);
        result.write_into(&mut self.trace_states);
```

Reply:
I think the above is fine. Maybe I would have added an extra newline like so:
```Rust
let frame_size = trace_states.len();
let width = trace_states[0].len();

let mut result = vec![];
for i in 0..width {
    for row in trace_states.iter() {
        result.push(row[i]);
    }
}

debug_assert!(frame_size <= u8::MAX as usize);
self.trace_states.write_u8(frame_size as u8);
result.write_into(&mut self.trace_states);
```
But, in my opinion, the difference is marginal at best.